### PR TITLE
feat: auto-rename git branch after first user prompt using Haiku

### LINKED ISSUE #123: feat: auto-rename git branch after first user prompt using Haiku
## Summary

After a workspace is created and the user sends their first prompt, use Claude Haiku to generate a descriptive git branch name that reflects the actual work — then rename the branch automatically. This replaces the random `claudette/quirky-daisy` style names with meaningful ones like `claudette/fix-login-timeout` or `claudette/add-csv-export`.

## Motivation

Today, workspaces get random adjective-plant names (e.g. `claudette/mighty-oak`) via the `NameGenerator`. These names are fun but carry no semantic meaning — you can't tell what work a branch contains without opening it. Once the user types their first prompt, we know *exactly* what the workspace is for and can generate a much better name.

## Current Flow

1. User creates workspace → `NameGenerator` produces a random slug (e.g. `quirky-daisy`)
2. Branch `claudette/quirky-daisy` is created via `git worktree add`
3. User sends first prompt → agent session starts
4. Branch name never changes

## Proposed Flow

1. User creates workspace → random name assigned as before (unchanged)
2. User sends first prompt → agent session starts as normal
3. **New**: After the first prompt is sent, fire a lightweight Haiku call in the background:
   - Input: the user's first prompt text
   - System prompt: generate a short, branch-safe slug (lowercase, hyphens, ≤50 chars) that summarizes the task
   - Output: e.g. `fix-login-timeout`
4. Rename the git branch: `claudette/quirky-daisy` → `claudette/fix-login-timeout`
5. Update the workspace record in SQLite (name column)
6. Emit a Tauri event so the frontend updates the sidebar and any other UI showing the workspace/branch name

### 1. Add a Haiku branch-naming function (`src/agent.rs` or new module)

- New async function, e.g. `generate_branch_name(prompt: &str) -> Result<String>`
- Invoke `claude` CLI with `--model haiku` (or the equivalent flag for Haiku) with a tightly scoped system prompt:
  > You are a git branch name generator. Given the user's task description, output a single branch-safe slug: lowercase ASCII, hyphens only, no spaces, max 50 characters. Output ONLY the slug, nothing else.
- Parse and sanitize the output (strip whitespace, enforce charset, truncate)
- Add unit tests for sanitization edge cases

### 2. Add a git branch rename helper (`src/git.rs`)

- New async function: `rename_branch(worktree_path: &Path, old_branch: &str, new_branch: &str) -> Result<()>`
- Shells out to `git branch -m <old> <new>` inside the worktree
- Handle error case where the new name conflicts with an existing branch (fall back to keeping the random name)
- Add tests

### 3. Add a workspace rename DB operation (`src/db.rs`)

- New function: `rename_workspace(conn: &Connection, workspace_id: i64, new_name: &str, new_branch: &str) -> Result<()>`
- Updates both the `name` and `branch` columns

### 4. Orchestrate in the chat command (`src-tauri/src/commands/chat.rs`)

- In `send_chat_message()`, after the first turn starts (when `session.turn_count == 0`):
  - Spawn a background Tokio task that:
    1. Calls `generate_branch_name(&prompt)`
    2. Calls `rename_branch(...)` 
    3. Calls `rename_workspace(...)` in the DB
    4. Emits a `workspace-renamed` Tauri event with `{ workspace_id, new_name, new_branch }`
  - This runs concurrently with the agent turn — does NOT block the chat
  - If any step fails, log a warning and keep the original name (non-fatal)

### 5. Handle the event in the frontend

- Listen for `workspace-renamed` event in the Zustand store or a hook
- Update the workspace entry in the store (name + branch)
- UI re-renders automatically via Zustand subscriptions (sidebar, status bar, etc.)

### 6. Add a user preference (optional, can defer)

- Setting in repo config or app preferences to disable auto-rename
- Some users may prefer the random names or want to name workspaces manually

## Edge Cases

- **User sends a vague prompt** (e.g. "help me"): Haiku should still produce something reasonable; worst case we keep the random name
- **Branch name collision**: If `claudette/fix-login-timeout` already exists, append a short suffix (e.g. `-2`) or keep the random name
- **Rename race condition**: The rename happens in a background task; if the user sends a second message very quickly, ensure the rename completes or is skipped gracefully
- **Workspace already renamed**: Guard against renaming on turn 2+ (only trigger when `turn_count == 0`)

## Testing

- Unit tests for branch name sanitization
- Unit tests for `rename_branch` git operation
- Integration test: create workspace → simulate first prompt → verify branch + DB updated
- Test that a failed rename doesn't break the chat flow

## Out of Scope

- Renaming after subsequent prompts (only first prompt triggers this)
- Renaming the worktree directory on disk (too disruptive; the directory can keep its original name)
- Pushing renamed branches to remote (user controls when to push)

### PR DESCRIPTION
**Pull request by doomspork**
## Summary

Closes #123

After a workspace is created and the user sends their first prompt, a background Haiku call generates a descriptive branch name (e.g. `claudette/fix-login-timeout`) and renames the branch automatically — replacing the random `claudette/quirky-daisy` style names.

- **`src/agent.rs`**: Added `generate_branch_name()` (one-shot Haiku CLI call) and `sanitize_branch_name()` to produce clean branch slugs
- **`src/git.rs`**: Added `rename_branch()` wrapping `git branch -m`
- **`src/db.rs`**: Added `rename_workspace()` to update name + branch in SQLite
- **`src-tauri/src/commands/chat.rs`**: Added `try_auto_rename()` orchestrator, spawned as a detached background task on first turn — does not block chat streaming
- **`src/ui/src/hooks/useAgentStream.ts`**: Added `workspace-renamed` event listener that updates the Zustand store for reactive sidebar updates

```mermaid
sequenceDiagram
    participant User
    participant ChatPanel
    participant send_chat_message
    participant Bridge Task
    participant Haiku Task
    participant DB
    participant Git
    participant Frontend

    User->>ChatPanel: First prompt
    ChatPanel->>send_chat_message: invoke()
    send_chat_message->>Bridge Task: tokio::spawn (event loop)
    Bridge Task->>Haiku Task: tokio::spawn (detached)
    Bridge Task->>Frontend: agent-stream events (not blocked)
    Haiku Task->>Haiku Task: claude --model haiku (generate slug)
    Haiku Task->>DB: rename_workspace()
    Haiku Task->>Git: git branch -m
    Haiku Task->>Frontend: emit "workspace-renamed"
    Frontend->>Frontend: updateWorkspace() → sidebar re-renders
```

## Complexity Notes

- **Cross-crate boundary**: `rusqlite` is a dep of the `claudette` lib crate but not the `src-tauri` binary crate. The unique constraint collision check uses string matching on the error message (`"UNIQUE constraint failed"`) rather than matching the SQLite error code directly.
- **Non-fatal by design**: All rename failures (Haiku timeout, name collision exhaustion, git error) log a warning and keep the original random name. The chat flow is never interrupted.

## Test Steps

1. `cargo test --all-features` — 171 tests pass, including 13 new tests:
   - `test_rename_branch`, `test_rename_branch_conflict` (git)
   - `test_rename_workspace`, `test_rename_workspace_unique_conflict` (db)
   - 9 `test_sanitize_*` tests (branch name sanitization edge cases)
2. `cargo clippy --workspace --all-targets` — zero warnings with `RUSTFLAGS="-Dwarnings"`
3. `cargo fmt --all --check` — clean
4. `bunx tsc --noEmit` — clean
5. Manual: create a workspace → send first prompt → observe branch name update in sidebar within a few seconds

## Checklist

- [x] Tests added/updated
- [ ] Documentation updated (if applicable)

### CODE REVIEW

File: src/db.rs [hunk 1]
```diff
@@ -484,6 +484,22 @@ impl Database {
...
+    pub fn rename_workspace(
+        &self,
+        id: &str,
+        new_name: &str,
+        new_branch_name: &str,
+    ) -> Result<(), rusqlite::Error> {
+        self.conn.execute(
+            "UPDATE workspaces SET name = ?1, branch_name = ?2 WHERE id = ?3",
+            params![new_name, new_branch_name, id],
+        )?;
```
[the remark this answers is not in the document]
Comment (doomspork) [reply src/db.rs]: Done — now checks `rows_affected != 1` and returns `Err(StatementChangedRows(0))` when the workspace ID doesn't exist. Added a test for this case too.

File: src/agent.rs [hunk 2]
```diff
@@ -403,6 +403,102 @@ pub async fn stop_agent(pid: u32) -> Result<(), String> {
...
+            prev_hyphen = false;
+        }
+    }
+
+    // Trim leading/trailing hyphens, truncate.
+    let trimmed = collapsed.trim_matches('-');
+    if trimmed.len() <= max_len {
+        return trimmed.to_string();
+    }
+    // Truncate at max_len but don't cut mid-word if there's a hyphen nearby.
```
[the remark this answers is not in the document]
Comment (doomspork) [reply src/agent.rs]: Done — updated the comment to accurately describe the behavior: truncate to max_len and drop trailing hyphens.

File: src/agent.rs [hunk 3]
```diff
@@ -403,6 +403,102 @@ pub async fn stop_agent(pid: u32) -> Result<(), String> {
...
+        "claude-haiku-4-5",
+        "--max-turns",
+        "1",
+        "--append-system-prompt",
+        "You are a git branch name generator. Given the user's task description, \
+         output a single branch-safe slug: lowercase ASCII letters, numbers, and \
+         hyphens only. No spaces, no quotes, no explanation. Max 40 characters. \
+         Output ONLY the slug.",
+        &truncated,
+    ]);
```
[the remark this answers is not in the document]
Comment (doomspork) [reply src/agent.rs:480]: Done — added `cmd.stdin(std::process::Stdio::null())` to prevent the background subprocess from inheriting stdin.

File: src-tauri/src/commands/chat.rs [hunk 4]
```diff
@@ -511,6 +541,75 @@ pub async fn load_completed_turns(
...
+                return;
+            }
+        };
+
+        match db.rename_workspace(ws_id, candidate, &new_branch) {
+            Ok(()) => {
+                // DB updated — now rename the git branch.
+                if let Err(e) = git::rename_branch(repo_path, old_branch, &new_branch).await {
+                    eprintln!("[rename] Git branch rename failed: {e} — rolling back DB");
+                    let _ = db.rename_workspace(ws_id, old_name, old_branch);
```
[the remark this answers is not in the document]
Comment (doomspork) [reply src-tauri/src/commands/chat.rs:587]: Done — now checks for "already exists" in the git error message and `continue`s to the next candidate instead of returning. Only non-retriable git errors cause an early return.

File: src-tauri/src/commands/chat.rs [hunk 5]
```diff
@@ -511,6 +543,85 @@ pub async fn load_completed_turns(
...
+
+                    // If the target branch already exists, fall back to the next
+                    // candidate just like we do for DB unique constraint collisions.
+                    if msg.contains("already exists") {
+                        eprintln!("[rename] Branch {new_branch:?} collides, trying next");
+                        continue;
+                    }
+                    return;
+                }
```
[the remark this answers is not in the document]
Comment (doomspork) [reply src-tauri/src/commands/chat.rs:596]: Addressed in #138 — now runs `git branch -m` from `worktree_path` instead of `repo_path`, and added a test that creates a worktree then renames the checked-out branch from within it.

File: src/git.rs [hunk 6]
```diff
@@ -262,6 +262,16 @@ pub async fn restore_to_commit(worktree_path: &str, commit_hash: &str) -> Result
...
 }
 
+/// Rename a branch. The worktree's HEAD follows automatically.
+pub async fn rename_branch(
+    repo_path: &str,
+    old_name: &str,
+    new_name: &str,
+) -> Result<(), GitError> {
+    run_git(repo_path, &["branch", "-m", old_name, new_name]).await?;
+    Ok(())
```
[the remark this answers is not in the document]
Comment (doomspork) [reply src/git.rs:272]: Addressed in #138 — added `test_rename_branch_checked_out_in_worktree` which creates a worktree, then verifies `rename_branch()` succeeds when called from the worktree path where the branch is checked out.
